### PR TITLE
Issue 13

### PR DESCRIPTION
**Pull request by o-smirnov**
Adds ``assign_based_on`` functionality, see https://github.com/o-smirnov/jovial/blob/master/image-boom-new.yml#L109 for a real-life example.

Merged in the slurm changes, and fixed failing test.

### CODE REVIEW

**Comment by o-smirnov** [comment]
@SpheMakh please note there's a tiny PR on scabha2 to support these changes: https://github.com/caracal-pipeline/scabha2/pull/13

File: stimela/kitchen/recipe.py [hunk 1]
```diff
@@ -356,18 +365,81 @@ def __post_init__ (self):
...
+
+    def validate_assignments(self, assign, assign_based_on, location):
+        # collect a list of all assignments
+        assignments = OrderedDict()
+        for key in assign:
+            assignments[key] = "assign"
+        for basevar, lookup_list in assign_based_on.items():
+            if not isinstance(lookup_list, Mapping):
+                raise RecipeValidationError(f"{location}.{assign_based_on}.{basevar}: mapping expected")
+            # for assign_list in lookup_list.values():
```
Comment (sphemakh) [inline stimela/kitchen/recipe.py:392]: Why not delete this old code. We have the git history if you need it again in the future

File: stimela/kitchen/recipe.py [hunk 2]
```diff
@@ -853,8 +941,9 @@ def loop_worker(inst, step, label, subst, count, iter_var):
                 time.sleep(1)
             results.get()
         else:
-            loop_pool = SerialPool(max_workers)
-            results = list(loop_pool.imap(loop_worker, *loop_args))
+            # loop_pool = SerialPool(max_workers)
```
Comment (sphemakh) [inline stimela/kitchen/recipe.py:944]: Ditto. Remove old code

**Review by SpheMakh — APPROVED** [review]
Looks good

**Pull request merged**